### DISCUSSION TITLE
[Question] How can I generate scripts without runtime (Pyarmor 8) (not in docs)

### DISCUSSION THREAD

**Discussion by pauloswear** [dashingsoft/pyarmor#1568] [General]
Hello, I was searching in docs something like **--no-runtime** from old Pyarmor-7. I assume I'm not familiar with the pyarmor-8 commands, I'm finding it confusing. Even after reading the documentation dozens of times, I still have several difficulties. However, I didn't find anything similar in the documentation about generate scripts without runtime. Can you provide us an example how to use same function in Pyarmor-8 ? 

Thanks.

**Comment by jondy** [comment]
Do you run `pyarmor gen -h` to list all options?
Do you check man page in the documentation to check all the available options in detail?

**Comment by jondy** [comment]
Sorry, this is not still documented, but `--no-runtime` still works.
You can run `pyarmor gen -h` to list all available options.
You also need use `--use-runtime PATH` to specified pre-build runtime package when generating scripts.

**Reply by pauloswear** [reply]
--no-runtime requires outer, not documented yet. It's documented now how to use with --use-runtime, this resolves when you want to share runtime packages, but no any no-runtime to generate scripts without this import, as old pyarmor-7.


new start script with --use-runtime:

> from pyarmor_runtime_vax import __pyarmor__

----

old start script with --no-runtime:

> __pyarmor__(b'bytecode')

**Comment by jondy** [comment]
I have add this guide now, please check
https://pyarmor.readthedocs.io/en/latest/tutorial/advanced.html#using-shared-runtime-package

But there is a bug for `--use-runtime` in Pyarmor, it will be fixed in next release 8.4.3

Or verify it by pre-release version
  https://pyarmor.dashingsoft.com/downloads/temp/pyarmor-8.4.3.zip   
e

**Reply by pauloswear** [reply]
INFO     Python 3.10.5
INFO     Pyarmor 8.4.3 (pro), *,*
INFO     Platform windows.x86_64
ERROR    --outer is required if using --no_runtime

> pyarmor gen --use-runtime C:\Users\Paulo\Desktop\launcher\runtime inicializar.py

**Reply by pauloswear** [reply]
Following: https://pyarmor.readthedocs.io/en/latest/tutorial/advanced.html#using-shared-runtime-package

1st) 
```
pyarmor gen runtime -O build/my_runtime1
Ok, it generated runtime package with my vax code. (Preffer to hide VAX if exists any option)
```

2nd) 
```
pyarmor gen --use-runtime build/my_runtime1 foo.py
**ERROR --outer is required if using --no_runtime**
```

**Comment by jondy** [comment]
Try it again:
https://pyarmor.dashingsoft.com/downloads/temp/pyarmor-8.4.3.zip 
I have uploaded a new one.

**Reply by pauloswear** [reply]
This worked for --use-runtime, but --no-runtime has same error

Any option for **--no-runtime** ?

```
C:\Users\Paulo\Desktop\testes>pyarmor gen --no-runtime foo.py
INFO     Python 3.10.5
INFO     Pyarmor 8.4.3 (pro), **, ***
INFO     Platform windows.x86_64
ERROR    --outer is required if using --no_runtime
```

**Comment by jondy** [comment]
You need use `--use-runtime` to specify one pre-built runtime package for non-outer license in Pyarmor 8.

**Reply by pauloswear** [reply]
Can you provide us --no-runtime usage or document --no-runtime usage?

**Comment by jondy** [comment]
generally need not `--no-runtime`. It's not show in `pyarmor -h` and only for internal use.

**Reply by pauloswear** [reply]
--use-runtime I need to specify runtime path, wich it's ok. But the problem is that I can't customize from whre python will import runtime..., or even betten, i can't use same runtime that main script started


Doc:

> But it need copy shared runtime package to dist path:
> 
> pyarmor_runtime_000000 need to replaced with real name
> ls build/my_runtime1/
> cp -a build/my_runime1/pyarmor_runtime_000000 dist/

Old --no-runtime do not use this import  "from pyarmor_runtime_000 import pyarmor", wich allow's create only script and import from any code that already have imported pyarmor runtime...

Using this way I can't create only the script and import from other scripts that already have started with runtime pkg

**Reply by pauloswear** [reply]
How to reproduce and why this feature is necessary (at least for me)

1) Create this structure:

```
C:/folder1
    main.py

C:/folder2
    foo.py
```

2) Put this inside main.py:

```
import sys
sys.path.append("C:/folder2")
import foo.py
```

3) Inside foo.py

`
print("I can run")
`

4) Generate runtime inside C:/folder:
`pyarmor gen runtime -O build/my_runtime1`

5) Obf main.py inside C:/folder: 
`pyarmor gen --use-runtime build/my_runtime1 main.py`

6) Usually obf with --no-runtime with old pyarmor-7, but now there's no option yet

**Comment by jondy** [comment]
Well, I don't check your example. But both pyarmor 8 and pyarmor 7, the obfuscated scripts are normal python scripts, and the runtime package is also normal python package.
I think you can check pyarmor 8 options to find the solution, in pyarmor 8, these options `-i`, `--prefix` are used to generate runtime package
